extensions/filesystem/vault-fs/src/main/java/org/eclipse/dataspaceconnector/core/security/fs/FsPrivateKeyResolver.java

```diff
@@ -64,6 +77,58 @@ public FsPrivateKeyResolver(String password, KeyStore keyStore) {
 
     @Override
     public <T> @Nullable T resolvePrivateKey(String id, Class<T> keyType) {
-        return keyType.cast(privateKeyCache.get(id));
+        var key = privateKeyCache.get(id);
+
+        if (key == null) {
+            return null;
+        }
+
+        var keyParser = getParser(keyType);
+
+        if (keyParser == null) {
+            monitor.debug("No KeyParser available for type " + keyType);
+            return keyType.cast(privateKeyCache.get(id));
+        }
+
+        return keyType.cast(keyParser.parse(toPemEncoded(key)));
+    }
+
+    @Override
+    public <T> void addParser(KeyParser<T> parser) {
+        parsers.add(parser);
+    }
+
+    @Override
+    public <T> void addParser(Class<T> forType, Function<String, T> parseFunction) {
+        var p = new KeyParser<T>() {
+
+            @Override
+            public boolean canParse(Class<?> keyType) {
+                return Objects.equals(keyType, forType);
+            }
+
+            @Override
+            public T parse(String encoded) {
+                return parseFunction.apply(encoded);
+            }
+        };
+        addParser(p);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> KeyParser<T> getParser(Class<T> keytype) {
+        return (KeyParser<T>) parsers.stream().filter(p -> p.canParse(keytype))
+                .findFirst().orElse(null);
+    }
+
+    private String toPemEncoded(PrivateKey key) {
```

Comment:
Because in FsVault we have ECKey and RSAPrivateKey which needs to be converted to PEM encoded values in order to be parsed and convert to PrivateKeyWrapper for example.

In VaultPrivateKeyResolver we directly have PEM encoded values stored in cloud vaults.

Reply:
Would it make sense to have both vault implementations store keys in the same format? No need to do this in this PR though.

Reply:
I think currently it will not be feasible to do that as in FsKeyResolver we store this as PrivateKey but in other implementations it is stored as secret which can be a key or any other secret value, so cannot encode it.